_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,223 @@
+
+1. Aggregation field (payment amount)  
+1. Grouping fields (beneficiary + payee)  
+1. Aggregation function (sum)  
+1. Window duration (1 week)  
+1. Limit (1 000 000)  
+1. Limit operator (greater)  
+
+Accordingly, we will use the following simple JSON format to define the aforementioned parameters:
+
+```json  
+{
+  "ruleId": 1,
+  "ruleState": "ACTIVE",
+  "groupingKeyNames": ["beneficiaryId", "payeeId"],
+  "aggregateFieldName": "paymentAmount",
+  "aggregatorFunctionType": "SUM",
+  "limitOperatorType": "GREATER",
+  "limit": 1000000,
+  "windowMinutes": 10080
+}
+```
+
+At this point it is important to understand that **`groupingKeyNames`** determine the actual physical grouping of events - all Transactions with the same values of specified parameters (e.g. _beneficiary #25 -> payee #12_ have to be aggregated in the same physical instance of the evaluating operator. Naturally, the process of distributing data in such a manner in Flink's API is realised by a `keyBy()` function.
+
+Most examples in Flink's `keyBy()`[documentation](https://ci.apache.org/projects/flink/flink-docs-stable/dev/api_concepts.html#define-keys-using-field-expressions) use a hard-coded `KeySelector`, which extracts specific fixed events' fields. However, in order to support desired flexible parameters specification functionality, we have to extract them in a more dynamic fashion based on the specifications of the rules. For this, we will have to use one additional operator that prepares every event for dispatching to a correct aggregating instance.
+
+On a high level, our main processing pipeline looks like this:
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy(/* some key selector */);
+        .process(/* actual calculations and alerting */)
+```
+
+We have previously established that each rule defines a **`groupingKeyNames`** parameter. This parameter specifies which combination of fields will be used for the incoming events' grouping. Each rule might use an arbitrary combination of these fields. At the same time, every incoming event potentially needs to be evaluated against multiple rules. This implies that events might simultaneously need to be present at multiple parallel instances of evaluating operators which correspond to different rules and hence will need to be forked. Ensuring such events dispatching is the purpose of `DynamicKeyFunction()`.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-11-19-demo-fraud-detection/shuffle_function_1.png" width="800px" alt="Figure 3: Forking events with Dynamic Key Function"/>
+<br/>
+<i><small>Figure 3: Forking events with Dynamic Key Function</small></i>
+</center>
+<br/>
+
+ `DynamicKeyFunction` iterates over a set of defined rules and prepares every event to be processed by a `keyBy()` function by extracting required grouping keys:
```

Comment:
```suggestion
 `DynamicKeyFunction` iterates over a set of defined rules and prepares every event to be processed by a `keyBy()` function by extracting any required grouping keys:
```

Reply:
They are sort of concrete at this point, how about "the required grouping keys"?

Reply:
Makes sense. I have updated the suggestion with your comment.